### PR TITLE
Add DID DHT Method support

### PR DESCRIPTION
**Pull request by decentralgabe**
Follows https://tbd54566975.github.io/did-dht-method

- Support for type indexing and two key types (ed25519, secp256k1)
- Support for packet encoding and decoding with tests
- Publishing to the DHT and resolution from the DHT to come in a follow-up

### CODE REVIEW

File: dids/src/main/kotlin/web5/sdk/dids/ion/model/Models.kt
Comment (andresuribe87) [inline dids/src/main/kotlin/web5/sdk/dids/ion/model/Models.kt:27]: This seems to belong in another package, as it's not ion specific.

**Review by andresuribe87 — COMMENTED** [review]
Would suggest waiting until #96 is merged, as it changes the approach that's being taken across dids.

File: dids/build.gradle.kts
Comment (decentralgabe) [inline dids/build.gradle.kts:45]: was the only open source z-base-32 lib I could find with an acceptable license that worked for the JVM. I tried chad but chad failed and failed again, this works.

File: dids/src/main/kotlin/web5/sdk/dids/Models.kt
Comment (decentralgabe) [inline dids/src/main/kotlin/web5/sdk/dids/Models.kt:8]: @andresuribe87 we should consider killing this since it's present in the DID lib we use

File: dids/build.gradle.kts
Comment (KendallWeihe) [reply dids/build.gradle.kts:45]: wow what a name `m-m-m.mmm-binary` 😆

File: dids/src/test/kotlin/web5/sdk/dids/DidDhtTest.kt
Comment (KendallWeihe) [inline dids/src/test/kotlin/web5/sdk/dids/DidDhtTest.kt:92]:
```suggestion
  inner class DnsPacketTest {
```

File: dids/src/test/kotlin/web5/sdk/dids/DidDhtTest.kt
Comment (decentralgabe) [reply dids/src/test/kotlin/web5/sdk/dids/DidDhtTest.kt:95]: the DHT doesn't care about DNS at all. we are leveraging a project called [Pkarr](https://github.com/nuhvi/pkarr) which adds in the DNS semantics (worth reading!).

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (nitro-neal) [inline dids/src/main/kotlin/web5/sdk/dids/DidDht.kt:46]: wouldn't the default 80% case be to publish every time you make a new dht? Should we flip this default to true?

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (nitro-neal) [inline dids/src/main/kotlin/web5/sdk/dids/DidDht.kt:145]: Should we have a require that makes sure all parts of the servicesToAdd are there?
(What if we have a null type or service endpoint?)

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (nitro-neal) [inline dids/src/main/kotlin/web5/sdk/dids/DidDht.kt:143]:
```suggestion
      opts.servicesToAdd?.forEach { service ->
        requireNotNull(service.id) { "Service id cannot be null" }
        requireNotNull(service.type) { "Service type cannot be null" }
        requireNotNull(service.serviceEndpoint) { "Service serviceEndpoint cannot be null" }
      }
```

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (decentralgabe) [reply dids/src/main/kotlin/web5/sdk/dids/DidDht.kt:46]: yes, will add a note to do this when publishing is implemented

**Review by nitro-neal — APPROVED** [review]
LGTM after the build is fixed with the ZBase32 issue

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (mistermoe) [inline dids/src/main/kotlin/web5/sdk/dids/DidDht.kt:44]: thoughts on `verificationMethods` instead of `verificationMethodsToAdd`? i feel like the `toAdd` is implied because it's `create`

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (mistermoe) [inline dids/src/main/kotlin/web5/sdk/dids/DidDht.kt:45]: thoughts on `services` instead of `servicesToAdd`? i feel like the `toAdd` is implied because it's `create`

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (mistermoe) [inline dids/src/main/kotlin/web5/sdk/dids/DidDht.kt:110]: thoughts on using jwk thumbprint?

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (mistermoe) [inline dids/src/main/kotlin/web5/sdk/dids/DidDht.kt:320]: 🙈 apologies in advance for saying: `fromDnsPacket` to remain consistent with convention in this repo

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (mistermoe) [inline dids/src/main/kotlin/web5/sdk/dids/DidDht.kt:209]: added alternative suggestion for this function [in this PR](https://github.com/TBD54566975/web5-kt/pull/115)

**Review by mistermoe — COMMENTED** [review]
naisu @decentralgabe ! Added some comments / suggestions. Really excited about `did:dht`

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (decentralgabe) [reply dids/src/main/kotlin/web5/sdk/dids/DidDht.kt]: intentionally want to add a consistent first key to the DID document, thumbprint can be used for other keys in the doc

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (decentralgabe) [reply dids/src/main/kotlin/web5/sdk/dids/DidDht.kt]: I followed the naming convention in the rest of the project, you can see the ION create options are the same. I'm fine to rename here, but the others should be renamed as well.

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (decentralgabe) [reply dids/src/main/kotlin/web5/sdk/dids/DidDht.kt]: https://github.com/TBD54566975/web5-kt/pull/105#discussion_r1383759490

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (mistermoe) [reply dids/src/main/kotlin/web5/sdk/dids/DidDht.kt]: is the rationale behind why you want to do this documented somewhere? asking bc i know its going to come up over and over. would like to have a doc to point to

**Review by mistermoe — APPROVED** [review]

File: dids/src/main/kotlin/web5/sdk/dids/DidDht.kt
Comment (decentralgabe) [reply dids/src/main/kotlin/web5/sdk/dids/DidDht.kt]: no rationale (I can add) but it is specified here https://tbd54566975.github.io/did-dht-method/#create

the tl;dr on the rationale is:
- Have a guaranteed key to use (since it's the identity key, we can rely on it)
- Reduce size of the packet since 0 is shorter than a thumbprint

Comment (decentralgabe) [reply dids/src/main/kotlin/web5/sdk/dids/DidDht.kt]: https://github.com/TBD54566975/did-dht-method/issues/27#issuecomment-1795941823

Comment (mistermoe) [reply dids/src/main/kotlin/web5/sdk/dids/DidDht.kt]: awesome thanks!

**Pull request merged**
